common/src/main/java/org/apache/celeborn/common/network/protocol/TransportMessage.java

```diff
@@ -97,13 +99,11 @@ public <T extends GeneratedMessageV3> T getParsedPayload() throws InvalidProtoco
   }
 
   public ByteBuffer toByteBuffer() {
-    int totalBufferSize = payload.length + 4 + 4;
-    ByteBuffer buffer = ByteBuffer.allocate(totalBufferSize);
-    buffer.putInt(messageTypeValue);
-    buffer.putInt(payload.length);
-    buffer.put(payload);
-    buffer.flip();
-    return buffer;
+    CompositeByteBuf buffer = Unpooled.compositeBuffer();
+    buffer.writeInt(messageTypeValue);
+    buffer.writeInt(payload.length);
+    buffer.writeBytes(Unpooled.wrappedBuffer(payload));
```

Comment:
![image](https://github.com/apache/incubator-celeborn/assets/948245/5ab62dc1-011d-452e-8c6b-3e331e29e0b4)
Hi @bowenliang123 , I debugged into this code and find out that although `buffer.nioBuffer()` avoids memory copy, `buffer.writeBytes(Unpooled.wrappedBuffer(payload));` involves memory copy. And the logic becomes more complex.

Reply:
Hi, thanks for the feedback. May I ask which class is this screenshot for?

Reply:
We should change to use `addComponent(boolean, ByteBuf)` to append the wrapped buffer. But somehow it failed the CI tests.

Reply:
I ran `HugeDataTest` and the snapshot is from `UnpooledHeapByteBuf#setBytes`

Reply:
![image](https://github.com/apache/incubator-celeborn/assets/1935105/3fe39e51-cd95-474e-90ac-2515757b4ca7)

Reply:
It seems that with the `CompositeByteBuf` (or `wrappedBuffer()`) , getting NIO buffer by calling the `nioBuffer` method always cause new byte buffer allocation and byte copying.

Reply:
If we keep the current code changes, while updating the PR purpose for avoid manually controlling offset in `TransparentMessasge`'s NIO buffer assembling, would you like to consider this PR? @waitinfuture